### PR TITLE
add state transitions to guide api calls

### PR DESCRIPTION
**Pull request by adammoody**

### CODE REVIEW

File: src/scr.c [hunk 1]
```diff
@@ -37,6 +37,19 @@
 #include "dtcmp.h"
 #endif /* HAVE_LIBDTCMP */
 
+/* define which state we're in for API calls, this is to help ensure
+ * users call SCR functions in the correct order */
+typedef enum {
+    SCR_STATE_UNINIT,     /* before init and after finalize */
+    SCR_STATE_INIT,       /* between init/finalize */
+    SCR_STATE_RESTART,    /* between start/complete restart */
+    SCR_STATE_CHECKPOINT, /* between start/complete checkpoint */
```
Comment (gonsie) [inline src/scr.c]: Why support a checkpoint state and an output state? This will go away in 2.0? right?

File: src/scr.c [hunk 2]
```diff
@@ -37,6 +37,19 @@
 #include "dtcmp.h"
 #endif /* HAVE_LIBDTCMP */
 
+/* define which state we're in for API calls, this is to help ensure
+ * users call SCR functions in the correct order */
+typedef enum {
+    SCR_STATE_UNINIT,     /* before init and after finalize */
+    SCR_STATE_INIT,       /* between init/finalize */
```
Comment (gonsie) [inline src/scr.c]: I'm not sure I like the name "INIT" here because the states are then not mutually exclusive. Perhaps "IDLE"? or some other term.

Comment (adammoody) [inline src/scr.c]: IDLE sounds good.  I'll switch to that.

File: src/scr.c (hunk 1 above)
Comment (adammoody) [inline src/scr.c]: We had talked about keeping Checkpoint around as a convenience and/or backwards compatibility.  I'll leave this state in here until we remove the API calls.

**Comment by adammoody** [comment]
I'm fine with further discussions, but it may help to express the intent.  My main goal with this is to help users when they start integrating API calls.  It turns out that you can code in the wrong sequence, and things just happen to work by chance right now (e.g., calling route_file outside of a start/complete pair sometimes works).  That's a bug on the user's part in their use of the API, but they won't notice because it works by accident.  Eventually we may change the library so that it doesn't work anymore, and then they will suddenly be broken.  The goal with this is to help them detect incorrect API usage early in the integration stage.

**Pull request merged**